src/esp/gfx/PBRShader.h

```diff
@@ -0,0 +1,348 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#ifndef ESP_GFX_PBRSHADER_H_
+#define ESP_GFX_PBRSHADER_H_
+
+#include <memory>
+#include <unordered_set>
+#include <vector>
+
+#include <Corrade/Containers/EnumSet.h>
+#include <Magnum/GL/AbstractShaderProgram.h>
+#include <Magnum/Math/Matrix4.h>
+#include <Magnum/Trade/LightData.h>
+
+#include "esp/core/esp.h"
+
+namespace esp {
+
+namespace gfx {
+
+class PBRShader : public Magnum::GL::AbstractShaderProgram {
+ public:
+  /**
+   * @brief vertex positions
+   */
+  typedef Magnum::GL::Attribute<0, Magnum::Vector3> Position;
+
+  /**
+   * @brief normal direction
+   */
+  typedef Magnum::GL::Attribute<1, Magnum::Vector3> Normal;
+
+  /**
+   * @brief 2D texture coordinates
+   *
+   * Used only if at least one of
+   * @ref Flag::AlbedoTexture, @ref Flag::NormalTexture and
+   * @ref Flag::RoughnessTexture @ref Flag::MetallicTexture is set.
+   */
+  typedef Magnum::GL::Attribute<2, Magnum::Vector2> TextureCoordinates;
+
+  /**
+   * @brief Tangent direction with the fourth component indicating the handness.
+   *
+   * T = Tangent, B = BiTangent, N = Normal
+   *
+   * 1.0 means T, B, N form a right-handed coordinate;
+   * -1.0 means T, B, N form a left-handed coordinate;
+   *
+   * Used only if @ref Flag::NormalTexture is set.
+   */
+  typedef Magnum::GL::Attribute<3, Magnum::Vector4> Tangent4;
+
+  enum : Magnum::UnsignedInt {
+    /**
+     * Color shader output. @ref shaders-generic "Generic output",
+     * present always. Expects three- or four-component floating-point
+     * or normalized buffer attachment.
+     */
+    ColorOutput = 0,
+
+#ifndef MAGNUM_TARGET_GLES2
+    /**
+     * Object ID shader output. @ref shaders-generic "Generic output",
+     * present only if @ref Flag::ObjectId is set. Expects a
+     * single-component unsigned integral attachment. Writes the value
+     * set in @ref setObjectId() there.
+     * @requires_gl30 Extension @gl_extension{EXT,texture_integer}
+     * @requires_gles30 Object ID output requires integer support in
+     *      shaders, which is not available in OpenGL ES 2.0 or WebGL
+     *      1.0.
```

Comment:
These `@requires_gl` are magnum-specific and will break doc build here, remove them.